bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java

```diff
@@ -0,0 +1,199 @@
+    private @Nullable BufferedReader tcpInput = null;
+
+    private @Nullable IntesisBoxChangeListener changeListener;
+
+    private boolean connected = false;
+
+    public IntesisBoxSocketApi(final String ipAddress, final int port) {
+        this.ipAddress = ipAddress;
+        this.port = port;
+    }
+
+    private class IntesisSocket {
+        final Socket socket;
+
+        public IntesisSocket() throws UnknownHostException, IOException {
+            socket = new Socket();
+            SocketAddress tcpSocketAddress = new InetSocketAddress(ipAddress, port);
+            socket.connect(tcpSocketAddress);
+        }
+
+        public void close() throws IOException {
+            socket.close();
+        }
+    }
+
+    public void openConnection() throws IOException {
+        closeConnection();
+
+        IntesisBoxChangeListener listener = this.changeListener;
+        IntesisSocket localSocket = tcpSocket = new IntesisSocket();
+        tcpOutput = new OutputStreamWriter(localSocket.socket.getOutputStream(), "US-ASCII");
+        tcpInput = new BufferedReader(new InputStreamReader(localSocket.socket.getInputStream()));
+
+        Thread tcpListener = new Thread(new TCPListener());
+        tcpListener.start();
+
+        setConnected(true);
+        if (listener != null) {
+            listener.connectionStatusChanged(ThingStatus.ONLINE, null);
+        }
+    }
+
+    public void closeConnection() {
+        try {
+            if (tcpSocket != null) {
+                tcpSocket.close();
+                tcpSocket = null;
+            }
+            if (tcpInput != null) {
+                tcpInput.close();
+                tcpInput = null;
+            }
+            if (tcpOutput != null) {
+                tcpOutput.close();
+                tcpOutput = null;
+            }
+            setConnected(false);
+        } catch (IOException ioException) {
+            logger.debug("closeConnection(): Unable to close connection - {}", ioException.getMessage());
+        } catch (Exception exception) {
+            logger.debug("closeConnection(): Error closing connection - {}", exception.getMessage());
+        }
+    }
+
+    private class TCPListener implements Runnable {
+
+        /**
+         * Run method. Runs the MessageListener thread
+         */
+        @Override
+        public void run() {
+            while (isConnected()) {
+                String message = read();
+                readMessage(message);
+            }
+        }
+    }
+
+    public void addIntesisBoxChangeListener(IntesisBoxChangeListener listener) {
+        if (this.changeListener == null) {
+            this.changeListener = listener;
+        }
+    }
+
+    private void write(String data) {
+        IntesisBoxChangeListener listener = this.changeListener;
+        try {
+            if (tcpOutput != null) {
+                tcpOutput.write(data);
+            }
+            if (tcpOutput != null) {
```

Comment:
You shouldn't need to check if `tcpOutput != null` twice.
```suggestion
```

Reply:
If not checked again, Eclipse gives me a potential null pointer warning.

Reply:
Done